README
== Description
  A drop-in replacement for the current Pathname class.

== Prerequisites
  * facade
  * ffi (Windows only)
  * test-unit (testing only)
   
== Installation

  gem install pathname2

== Synopsis
  require 'pathname2'

  # Unix
  path1 = Pathname.new("/foo/bar/baz")
  path2 = Pathname.new("../zap")

  path1 + path2 # "/foo/bar/zap"
  path1 / path2 # "/foo/bar/zap" (same as +)
  path1.exists? # Does this path exist?
  path1.dirname # "/foo/bar"
  path1.to_a    # ['foo','bar','baz']

  # Windows
  path1 = Pathname.new("C:/foo/bar/baz")
  path2 = Pathname.new("../zap")

  path1 + path2 # "C:\\foo\\bar\\zap"
  path1.root    # "C:\\"
  path1.to_a    # ['C:','foo','bar','baz']

== Windows Notes
  All forward slashes are converted to backslashes for Pathname objects.

== Differences between Unix and Windows
  If your pathname consists solely of ".", or "..", the return
  value for Pathname#clean will be different. On Win32, "\\" is returned,
  while on Unix "." is returned.  I consider this an extreme edge case and
  will not worry myself with it.

== Differences between Pathname in the standard library and this version
  * It is a subclass of String (and thus, mixes in Enumerable).
  * It has sensical to_a and root instance methods.
  * It works on Windows and Unix.  The current implementation does not work
    with Windows path names very well, and not at all when it comes to UNC
    paths.
  * The Pathname#cleanpath method works differently - it always returns
    a canonical pathname.  In addition, there is no special consideration
    for symlinks (yet), though I'm not sure it warrants it.
  * The Pathname#+ method auto cleans.
  * It uses a facade for all File and Dir methods, as well as most FileUtils
    methods.
  * Pathname#clean works slightly differently.  In the stdlib version,
    Pathname#clean("../a") returns "../a".  In this version, it returns "a".
    This affects other methods, such as Pathname#relative_path_from.
  * Accepts file urls and converts them to paths automatically, e.g.
    file:///foo%20bar/baz becomes '/foo/bar/baz'.
  * Adds a Kernel level +pn+ method as a shortcut.
  * Allows you to add paths together with the '/' operator.
     
== Method Priority
  Because there is some overlap in method names between File, Dir, and
  FileUtils, the priority is as follows:
   
  * File
  * Dir
  * FileUtils
   
  In other words, whichever of these defines a given method first is the
  method that is used by the pathname2 library.
   
== Known Issues
  On MS Windows, some methods may not work on pathnames greater than 260
  characters because of internal function limitations.
   
  Any issues you find should be reported on the project page at
  https://github.com/djberg96/pathname2

== Future Plans
  Suggestions welcome.
   
== License
  Apache-2.0
   
== Copyright
  (C) 2003-2020 Daniel J. Berger
  All rights reserved.

== Warranty
  This library is provided "as is" and without any express or
  implied warranties, including, without limitation, the implied
  warranties of merchantability and fitness for a particular purpose.

== Author
  Daniel J. Berger


CHANGES
== 1.8.4 - 19-Jan-2021
* Cleaned up the Rubocop warnings.
* Added a Gemfile.
* Some minor updates to the gemspec and Rakefile.

== 1.8.3 - 30-Aug-2020
* Updated the tests to match the version. Thanks go to Cédric Boutillier for the spot.
* Fixed a redefinition warning for the :children method.
* Fixed a Fixnum deprecation warning.

== 1.8.2 - 21-Jul-2020
* Added a LICENSE file as required by the Apache-2.0 license.

== 1.8.1 - 31-Jan-2019
* The VERSION constant is now frozen.
* Added metadata to the gemspec.
* Fixed missing hyphen in license name.
* Updated cert, should be good for about 10 years.

== 1.8.0 - 19-Jun-2016
* Changed license to Apache-2.0.
* Some cleanup and updates to the Rakefile and benchmarks.
* Refactored some realpath tests so they're not touching system files like
  /dev/stdin any more. Thanks go to Michael R. Crusoe for pointing out the
  potential pitfalls of doing that.

== 1.7.4 - 7-Sep-2015
* This gem is now signed.
* Rakefile now assumes Rubygems 2.x for some tasks.

== 1.7.3 - 24-Apr-2014
* The basename and dirname methods now return Pathname objects.

== 1.7.2 - 23-Apr-2014
* The join and expand_path methods now return Pathname objects.

== 1.7.1 - 28-Mar-2014
* Updated gemspec for Windows vs Unix.
* Updated README and MANIFEST.

== 1.7.0 - 28-Mar-2014
* Windows now uses FFI internally instead of win32-api.
* Added a custom pretty_print method for the pp library.
* Reorganized the test suite for Windows, and added many
  more test tasks to the Rakefile.

== 1.6.5 - 19-Sep-2011
* Added the String#to_path instance method. This returns a Pathname
  object.
* The architecture for the Windows gem is now "universal", and some
  gemspec dependencies were simplified.

== 1.6.4 - 20-Jan-2011
* Explicitly remove the Pathname const if it is already defined in order
  to avoid a superclass mismatch error. This library assumes that if you
  require pathname2, you want my version of the Pathname class.
* Updated URI handling for Ruby 1.9.x.
* Added the Pathname() method, a synonym for Pathname.new.
* Some Rakefile and gemspec tweaks.
* Some updates to the test suite, including some specifically for Windows 7.

== 1.6.3 - 2-Oct-2009
* Updated Windows platform handling code to include mingw and cygwin.
* Added the :gem rake task.
* Minor gemspec updates.
* Some minor test suite updates.

== 1.6.2 - 4-Aug-2009
* Now compatible with Ruby 1.9.x.
* License changed to Artistic 2.0.
* Added the Pathname.pwd method as a synonym for Pathname.new(Dir.pwd).
* Modified Pathname#dirname so that you can specify a level that indicates
  how many levels up you want to retrieve. For example, if your path was
  '/usr/local/bin', then path.dirname(2) would return '/usr'.
* Now compatible with Ruby 1.9.x.
* Bumped required version of the facade library to 1.0.4.

== 1.6.1 - 8-Nov-2008
* Added the Pathname#[] method, which accepts an index, an index plus a
  length, or a range, and returns appropriate the path element(s).
* Refactoring the platform checking in the test suite to use rbconfig instead
  of RUBY_PLATFORM.
* More inline documentation examples.

== 1.6.0 - 13-July-2008
* The facade for ftools (and ftools itself) has been removed. The ftools
  library is deprecated in favor of FileUtils.
* PathnameError is now Pathname::Error.
* Bug fix for Pathname#relative_path_from for MS Windows. Thanks go to an
  anonymous user for the spot.
* Fixed a bug where frozen strings would raise an error on MS Windows.
* The code is now -w clean.
* Removed the C version as part of the release because it was just too
  difficult to maintain both versions. The C code remains in CVS, however.
* Changed platform checking to use rbconfig instead of RUBY_PLATFORM to avoid
  potential issues with other Ruby implementation.

== 1.5.2 - 9-Mar-2007
* Bug fix for the Pathname#realpath method where it was not handling recursive
  symlinks. The C version was also fixed, but it only affected platforms that
  don't have the realpath() function.
* Added a test for recursive symlinks (for Solaris, anyway).
* Updated the docs for Pathname#realpath.
* Minor speed enhancements for the C version and the elimination of one
  (potential) segfault.
* Added a 'Future Plans' section to the README.
* Added a Rakefile. You can now build, clean, and test and install (both the
  pure Ruby and C versions).

== 1.5.1 - 28-Aug-2006
* Added the Kernel#pn method as a shortcut for Pathname.new.
* The Pathname#readlink now properly handles symbolic links.  The 'fix'
  from 1.4.4 did not work.
* The C extension uses your system's realpath() function for the
  Pathname#readlink method if it has one.
* Added the '/' alias for '+'.  So, p1 / p2 is the same as p1 + p2.
* The windows-pr package is now required on MS Windows.

== 1.5.0 - 17-Apr-2006
* Better subclass handling, in that some methods that previously returned
  hardcoded Pathname.new now return self.class.new.
* File URL's are now handled on Unix as well (using the 'uri' package).
* Some comment changes/clarifications.
* Added the PathnameError class to the pure Ruby version.

== 1.4.4 - 23-Mar-2006
* If the path is a symbolic link the Pathname#realpath method
  now returns the absolute path of that link, i.e. the result
  of File.readlink (as a Pathname object).

== 1.4.3 - 3-Mar-2006
* Added the Pathname#realpath method.

== 1.4.2 - 22-Feb-2006
* Fixed the Pathname#relative_path_from method for Windows.  This really only
  affected edge cases that you probably never saw anyway.
* Added corresponding tests for Windows.

== 1.4.1 - 18-Feb-2006
* Added the Pathname#parent method.
* Added the Pathname#relative_path_from method.
* Bug fix for Pathname#pstrip on *nix.
* Corresponding test suite additions.

== 1.4.0 - 19-Dec-2005
* Added destructive and non-destructive methods for some methods - pstrip,
  pstrip!, undecorate, undecorate!, clean and clean!.
* Added a C extension version of this package.  You can use the C version

benchmarks/pathname_vs_pathname2.rb
#####################################################################
# pathname_vs_pathname2.rb
#
# Benchmark suite for all methods of the Pathname class, excluding
# the facade methods.
#
# Use the Rake tasks to run this benchmark:
#
# => rake benchmark to run the pure Ruby benchmark.
#####################################################################
require 'benchmark'
require 'rbconfig'

MAX = 100000

['pathname', 'pathname2'].each do |lib|
  require lib
  puts "\nLIB: #{lib}\n"

  if RbConfig::CONFIG['host_os'] =~ /mingw|mswin/i
     path1 = Pathname.new("C:\\Program Files\\Windows NT")
     path2 = Pathname.new("Accessories")
     path3 = Pathname.new("C:\\Program Files\\..\\.\\Windows NT")
  else
     path1 = Pathname.new("/usr/local")
     path2 = Pathname.new("bin")
     path3 = Pathname.new("/usr/../local/./bin")
     path4 = Pathname.new("/dev/stdin")
  end

  Benchmark.bm(25) do |bench|
    bench.report("Pathname.new(path)"){
      MAX.times{ Pathname.new("/usr/local/bin") }
    }

    bench.report("Pathname#+(Pathname)"){
      MAX.times{ path1 + path2 }
    }

    bench.report("Pathname#+(String)"){
      MAX.times{ path1 + path2 }
    }

    bench.report("Pathname#children"){
      MAX.times{ path1.children }
    }

    bench.report("Pathname#descend"){
      MAX.times{ path1.descend{} }
    }

    bench.report("Pathname#ascend"){
      MAX.times{ path1.ascend{} }
    }

    bench.report("Pathname#root?"){
      MAX.times{ path1.root? }
    }

    bench.report("Pathname#<=>"){
      MAX.times{ path1 <=> path2 }
    }

    bench.report("Pathname#absolute?"){
      MAX.times{ path1.absolute? }
    }

    bench.report("Pathname#relative?"){
      MAX.times{ path1.relative? }
    }

    bench.report("Pathname#cleanptah"){
      MAX.times{ path3.cleanpath }
    }
  end
end


benchmarks/bench_pathname.rb
#####################################################################
# bench_pathname.rb
#
# Benchmark suite for all methods of the Pathname class, excluding
# the facade methods.
#
# Use the Rake tasks to run this benchmark:
#
# => rake benchmark to run the pure Ruby benchmark.
#####################################################################
require 'benchmark'
require 'pathname2'
require 'rbconfig'

if RbConfig::CONFIG['host_os'] =~ /mingw|mswin/i
   path1 = Pathname.new("C:\\Program Files\\Windows NT")
   path2 = Pathname.new("Accessories")
   path3 = Pathname.new("C:\\Program Files\\..\\.\\Windows NT")
else
   path1 = Pathname.new("/usr/local")
   path2 = Pathname.new("bin")
   path3 = Pathname.new("/usr/../local/./bin")
   path4 = Pathname.new("/dev/stdin")
end

MAX = 100000

Benchmark.bm(25) do |bench|
  bench.report("Pathname.new(path)"){
    MAX.times{ Pathname.new("/usr/local/bin") }
  }

  bench.report("Pathname#+(Pathname)"){
    MAX.times{ path1 + path2 }
  }

  bench.report("Pathname#+(String)"){
    MAX.times{ path1 + path2 }
  }

  bench.report("Pathname#children"){
    MAX.times{ path1.children }
  }

  bench.report("Pathname#pstrip"){
    MAX.times{ path1.pstrip }
  }

  bench.report("Pathname#pstrip!"){
    MAX.times{ path1.pstrip! }
  }

  bench.report("Pathname#to_a"){
    MAX.times{ path1.to_a }
  }

  bench.report("Pathname#descend"){
    MAX.times{ path1.descend{} }
  }

  bench.report("Pathname#ascend"){
    MAX.times{ path1.ascend{} }
  }

  bench.report("Pathname#root"){
    MAX.times{ path1.root }
  }

  bench.report("Pathname#root?"){
    MAX.times{ path1.root? }
  }

  bench.report("Pathname#<=>"){
    MAX.times{ path1 <=> path2 }
  }

  bench.report("Pathname#absolute?"){
    MAX.times{ path1.absolute? }
  }

  bench.report("Pathname#relative?"){
    MAX.times{ path1.relative? }
  }

  bench.report("Pathname#clean"){
    MAX.times{ path3.clean }
  }

  bench.report("Pathname#clean!"){
    MAX.times{ path3.clean! }
  }

  # Platform specific tests
  if RbConfig::CONFIG['host_os'] =~ /mingw|mswin/i
    bench.report("Pathname.new(file_url)"){
      MAX.times{ Pathname.new("file:///C:/usr/local/bin") }
    }

    bench.report("Pathname#drive_number"){
      MAX.times{ path1.drive_number }
    }

    bench.report("Pathname#unc?"){
      MAX.times{ path1.unc? }
    }

    bench.report("Pathname#undecorate"){
      MAX.times{ path1.undecorate }
    }

    bench.report("Pathname#undecorate!"){
      MAX.times{ path1.undecorate! }
    }

    bench.report("Pathname#short_path"){
      MAX.times{ path1.short_path }
    }

    bench.report("Pathname#long_path"){
      MAX.times{ path1.long_path }
    }
  else
    bench.report("Pathname#realpath"){
      MAX.times{ path4.realpath }
    }
  end
end
